### PR TITLE
Speed up PASSES_RANDOMLY via structured RNG

### PR DESCRIPTION
**Pull request by mrgriffin**
Replaces `Random()` calls that are tested by a `PASSES_RANDOMLY` battle test to be either `RandomUniform` or `RandomWeighted` (depending on use). This allows the battle test runner to run the battle the minimum number of times to exhaustively test the outcomes.

See `include/random.h`.

Incorporates #2696 to write tests against the non-test implementations of `RandomUniform` and `RandomWeighted`.

### CODE REVIEW

File: src/battle_util.c [hunk 1]
```diff
@@ -3481,7 +3481,7 @@ u8 AtkCanceller_UnableToUseMove(void)
         case CANCELLER_FROZEN: // check being frozen
             if (gBattleMons[gBattlerAttacker].status1 & STATUS1_FREEZE && !(gBattleMoves[gCurrentMove].flags & FLAG_THAW_USER))
             {
-                if (Random() % 5)
+                if (RandomWeighted(WRNG_FROZEN, 1, 4))
```
Comment (SBird1337) [inline src/battle_util.c:3484]: Generally I think these changes will make the code a lot more readable, but I'm having a bit of trouble parsing those `RandomWeighted` statements.

Would it be more readable if we wrote
```
if (RandomWeighted(WEIGHTED_RNG_FROZEN, 20, 80))
```
or even had something like `RandomPercent(WEIGHTED_RNG_FROZEN, 80)`?

File: test/ability_compound_eyes.c [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 SINGLE_BATTLE_TEST("Compound Eyes raises accuracy")
 {
-    PASSES_RANDOMLY(91, 100);
+    PASSES_RANDOMLY(91, 100, WRNG_ACCURACY);
```
Comment (SBird1337) [inline test/ability_compound_eyes.c:6]: Somehow this is easier for me to parse as "X out of Y", but maybe we could consider percentage here as well?
```
PASSES_RANDOMLY(20, WEIGHTED_RNG_ACCURACY);
```
would say "20% of the tests need to pass.

File: test/test_battle.h [hunk 3]
```diff
@@ -65,8 +65,9 @@
  * single turn. MOVE causes the player to use Stun Spore and adds the
  * move to the Pokémon's moveset if an explicit Moves was not specified.
  * Pokémon that are not mentioned in a TURN use Celebrate.
- * The test runner attempts to rig the RNG so that the first move used
- * in a turn does not miss and activates its secondary effects (if any).
+ * The test runner rigs the RNG so that unless otherwise specified,
+ * moves always hit, never critical hit, always activate their secondary
+ * effects, and always roll the same damage modifier.
```
Comment (SBird1337) [inline test/test_battle.h:70]: It is not super clear to me now how I would specify that I want either randomness (idk if thats a requirement) or a specific mechanic to still work, e.g. how would I test against crits? Do I need to specify a `PASSES_RANDOMLY` with a tag?

**Review by SBird1337 — COMMENTED** [review]
Mainly some thoughts of mine, feel free to turn them down if you don't like it. (Did not mark it as change request)

Great job on updating the RNG. It will probably also improve the readability of the `Random` calls in general, which is pretty cool :)

**Comment by Jaizu** [comment]
Please fix conflicts :D

**Comment by mrgriffin** [comment]
Fixed conflicts. There's a failing test but I don't have time to work out what's gone on there right now. It's weird, one of the Defog ones that doesn't use randomness is now being flagged as having an illegal move :thinking:

**Comment by AsparagusEduardo** [comment]
> one of the Defog ones that doesn't use randomness is now being flagged as having an illegal move 🤔

I have that same issue in #2827 :O

**Review by SBird1337 — APPROVED** [review]

**Pull request merged**